### ISSUE TITLE
Change the default batched_reduce_size of search requests

### ISSUE THREAD

**Issue by jimczi** [elastic/elasticsearch#51857]
Today we execute a partial reduce of search requests after we buffered at least `512` shard search results. The default, users can change this value with `batched_reduce_size=N`,  seems quite high and can cause memory issue for queries that target a large amount of shards. We also want to use the partial reduce to speed up the search on subsequent search shard request (#51852) but users won't see the benefit unless they reduce the batched reduce size explicitly. Partial (and final) reduce are usually very fast so I am opening this issue to use a sane default that could save memories on coordinating node and speed up sorted queries on time-based indices (queries that can target a lot of shards). We have plenty of options so here's a non-exhaustive list:

1. Reduce the default to 5-10, that could slightly increase the overall latency but the benefit are non-negligible on time-based indices.
2. Reduce the default only for queries that can use the partial reduce to speed up subsequent shard search (sorted queries on time based indices). 
3. Change the threshold to check the size of the buffered results rather than the absolute number of shards. We could also trigger partial reduce based on inactivity (if no more shard responded in the last N seconds/minutes). 
4. Keep the default as it is :(

I am curious to hear your thoughts on these options.

**Comment by polyfractal** [comment]
Wow yeah, 512 is high :)  

I like the idea of triggering a reduce based on a timeout, but that seems potentially tricky to get right.  And as you said, reductions _should_ be pretty fast, so a lowish 5-50 default seems reasonable to me.

**Comment by jpountz** [comment]
I like the idea of just decreasing it to 5 or 10. It's very simple and most requests that hit a single index will keep doing a single reduction. There are some downsides too but the benefits outweigh downsides in my opinion.

**Comment by jimczi** [comment]
We've made several progress on this issue recently so apologies for not updating earlier.
 
While discussing the proposed change offline, we've decided to start with benchmarks to evaluate the effect of changing the default value.
We've started with the  `terms` aggregation and opened https://github.com/elastic/elasticsearch/pull/53216 to fix a performance bug.
In the meantime @nik9000 has changed the [incremental reduce](https://github.com/elastic/elasticsearch/pull/53793) to lazily deserialize shard result aggregations on partial reduce. 
This has two main advantages:
- We can now evaluate the memory cost of keeping shard aggregation results in memory more precisely. This is important if we want to allow partial reduce to be executed when a memory threshold is reached.
- We create the java representation of the aggregation tree only during the partial reduce. This makes these java objects short-lived which should help the overall garbage collection in the node.

In order to be able to effectively reduce the default value without impacting users, we think that more work is needed:

- [ ] Perform real merge-sort reduce on terms aggregations. This should reduce the cost of running incremental reduce.
- [ ] Provides micro-benchmark for date_histograms and histograms with different `batched_reduce_size`.

Once these tasks are completed we will resume the discussion with the additional informations in order to take a decision based on real measurements.

**Comment by nik9000** [comment]
I wonder if the number should be different for searches with aggs vs searches without them. Aggs *are* the most expensive thing to reduce but they are also the biggest thing to keep around.

**Comment by jpountz** [comment]
My inclination would be to try hard to have the same number all the time. I've been hit by moving parts like that a couple times when digging performance issues.

**Comment by hackerwin7** [comment]
Is there any plan to support parallel reduce on coordinator.
IMO, If a search request's shard requests count is extremely huge, and batched reduce size is relatively low. the subsequent shard query result maybe block on synchronized `consumeInternal()`.
how about support parallel reduce on this ?

**Comment by hackerwin7** [comment]
I test a huge case that if batched size = 5, the total query took is extremely high compare to default batched size parameters

**Comment by jimczi** [comment]
> how about support parallel reduce on this ?

The plan is to move to a thread pool in order to limit the number of partial reduce we perform in parallel. So the synchronization will be removed but we'll continue to limit the number of partial reduce executed in parallel for a single search request to 1.

>  if batched size = 5, the query's took is extremely high compare to default batched size parameters

We're making improvements to the partial reduce but this is expected to be slower. We didn't settle the default yet and 5 might be too small. We'll test the performance before doing the change but first we need to cleanup the synchronization and move the operation into a thread pool. Maybe you can share your results for comparison ?

**Comment by hackerwin7** [comment]
@jimczi 
I have implement a path of parallel reduce for master branch.
here is my test results:

## Mem stats

JVM | default | batched=5 | parallel
-- | -- | -- | --
Eden | increase 8.4 G | Increase more than Eden capacity | Increase more than Eden capacity
Old | Increase 1.6 G | Increase 200 MB | Increase 6 G
YGC count | 0 | 3 | 15
FGC count | 0 | 0 | 0

## Search stats

Query | default | batched=5 | parallel
-- | -- | -- | --
Huge-aggs-query | 14.184 s | 142.341 s | 37.681 s

**Comment by hackerwin7** [comment]
> The plan is to move to a thread pool in order to limit the number of partial reduce we perform in parallel. So the synchronization will be removed but we'll continue to limit the number of partial reduce executed in parallel for a single search request to 1.

I have a similar patch for this.

**Comment by javanna** [comment]
With #121885 ,  each data node performs incremental reduction. For now `batched_reduce_size` is honored on both data node and coord node. A natural follow-up would be to decouple the two aspects that this parameters controls: 

- how often incremental results are made available to async searches - this could become non configurable: as soon as a data node replies with its batched response, the coord nod would make such results available
- 
- CPU vs memory trade-off around performing more or less frequent partial reductions - this should stay configurable but should it be applied to only the data node side and be expressed in terms of a hint to e.g. save memory by doing more frequent partial reductions?

**Comment by original-brownbear** [comment]
Just a side note, I'll get to this shortly in earnest: I think the batched_reduce_size will become just a UX knob shortly. I have a working prototype that (using the existing code for both and just removing layers really) just does streaming result merging. There shouldn't be any more tradeoffs here between CPU and memory in the near term and we can decide what makes sense from the user interface perspective. Maybe there it's not so much about a batch size but more about a time/the frequency of new results becoming visible?

**Comment by javanna** [comment]
> time/the frequency of new results becoming visible?

I think that should be fixed, and that's made easier by batched execution: every time a data node comes back with results. I wish that that side of the knob is no longer needed.

As for how often to perform partial reduction on data node, I also wish we could remove the knob, but I am not super sure we can.

**Comment by original-brownbear** [comment]
> As for how often to perform partial reduction on data node, I also wish we could remove the knob, but I am not super sure we can.

We actually can :) It's quite easy too, sorry I'm slow at opening a PR for this because of all the other things that are ongoing but the situation boils down to:

* reducing top hits is trivially cheap anyway, we can just do that on every result even if we feel the need to.

but far more importantly for this discussion, check out the `org.elasticsearch.search.aggregations.AggregatorReducer` interface. We currently set up a reducer per partial reduction step, pass e.g. 5 results into it a call `get` to get the intermediary result.
There is nothing whatsoever (I tried it in a dirty PR already, it works) from just passing each and every individual result as we receive it into this thing. Aside from some details on making the concurrency behind this perform well, this is quite the trivial change (and even the performance is somewhat irrelevant vs. the current solution no matter how basic it's done ... at least that's what profiling suggests).
=> the thing can be made a UX only topic. For top hits trivially so and without any overhead. For aggs, exposing intermediary results would however require retaining the old logic and doing the above suggested optimization on a batch-by-batch basis , thus taking a likely visible but limited performance hit.

That was pretty much my plan here, time permitting :)
